tutorials/application_design/c++11/subscriber.cxx

```diff
@@ -0,0 +1,254 @@
+//
+//  RTI grants Licensee a license to use, modify, compile, and create derivative
+//  works of the Software.  Licensee has the right to distribute object form
+//  only for use with RTI products.  The Software is provided "as is", with no
+//  warranty of any type, including any warranty for fitness for any purpose.
+//  RTI is under no obligation to maintain or support the Software.  RTI shall
+//  not be liable for any incidental or consequential damages arising out of the
+//  use or inability to use the software.
+//
+
+#include <iostream>
+#include <sstream>
+#include <thread>
+#include <unordered_map>
+
+#include <rti/rti.hpp>
+#include "VehicleModeling.hpp"
+#include "common.hpp"
+
+struct DashboardItem {
+    std::string vin;
+    bool is_historical;
+    std::vector<double> fuel_history;
+    int completed_routes;
+    rti::core::optional<Coord> current_destination;
+    std::vector<Coord> reached_destinations;
+};
+
+struct SubscriberDashboard {
+    explicit SubscriberDashboard(
+            dds::sub::DataReader<VehicleMetrics> metrics_reader,
+            dds::sub::DataReader<VehicleTransit> transit_reader)
+            : metrics_reader_(metrics_reader), transit_reader_(transit_reader)
+    {
+    }
+
+    void run();
+
+    friend std::string utils::to_string<>(const SubscriberDashboard &dashboard);
+
+private:
+    dds::sub::DataReader<VehicleMetrics> metrics_reader_;
+    dds::sub::DataReader<VehicleTransit> transit_reader_;
+    dds::core::cond::WaitSet waitset;
+    std::unordered_map<dds::core::InstanceHandle, DashboardItem>
+            dashboard_data_;
+
+    void display_app();
+    void metrics_app();
+    void transit_app();
+
+    std::vector<DashboardItem> online_vehicles()
+    {
+        std::vector<DashboardItem> online;
+        for (auto &item : dashboard_data_) {
+            if (!item.second.is_historical) {
+                online.push_back(item.second);
+            }
+        }
+        return online;
+    }
+
+    std::vector<DashboardItem> offline_vehicles()
+    {
+        std::vector<DashboardItem> offline;
+        for (auto &item : dashboard_data_) {
+            if (item.second.is_historical) {
+                offline.push_back(item.second);
+            }
+        }
+        return offline;
+    }
+};
+
+void SubscriberDashboard::run()
+{
+    std::mutex mutex;
+
+    dds::sub::cond::ReadCondition metrics_condition(
+            metrics_reader_,
+            dds::sub::status::DataState::any(),
+            [this]() { metrics_app(); });
+
+    dds::sub::cond::ReadCondition transit_condition(
+            transit_reader_,
+            dds::sub::status::DataState::any(),
+            [this]() { transit_app(); });
+
+    dds::core::cond::GuardCondition display_condition;
+    display_condition.extensions().handler(
+            [this, &display_condition]() { display_app(); });
+
+    std::thread display_thread([&display_condition, &mutex]() {
```

Comment:
Did you try calling `trigger_value(false)` inside the condition handler? That would remove the need for the mutex.

Otherwise I think it would be simpler to just use the thread to call `display_app()`, use the mutex to protect display/transit/metrics) and remove the GuardCondition.

Reply:
The main problem with having the thread execute `display_app` is that I would need to protect the concurrency access and modification of `dashboard_data_`.

I believe that I attempted to use `trigger_value` in the condition handler, but decided to keep it in the main loop to have the three handlers as simpler lambdas.

Reply:
I think calling trigger_value(false) in the handler makes the code simpler, as it doesn't require the mutex.